### ISSUE TITLE
Arabic to English and English to Arabic fine-tuning on WikiMatrix dataset

### ISSUE THREAD

**Issue by humamrashid** [pytorch/fairseq#3489]
Hi, I'm fine-tuning for an English to Arabic MT project and used the following tutorial (adapted to En->Ar) to get started:

[Fine-Tune NMT with mBART](https://tmramalho.github.io/science/2020/06/10/fine-tune-neural-translation-models-with-mBART/)

I used the WikiMatrix dataset for Ar-En and was able to produce checkpoints but the actual translations seem to work in reverse, i.e., when English is the source and Arabic is the target, I get no translations from English to Arabic but do get translations from Arabic to English. I used the following command for the fine-tuninng (adapted from the [official instructions](https://github.com/pytorch/fairseq/tree/master/examples/mbart)):

`fairseq-train <path_to_data> \
  --encoder-normalize-before --decoder-normalize-before \
  --arch mbart_large --layernorm-embedding \
  --task translation_from_pretrained_bart \
  --source-lang en_XX --target-lang ar_AR \
  --criterion label_smoothed_cross_entropy --label-smoothing 0.2 \
  --optimizer adam --adam-eps 1e-06 --adam-betas '(0.9, 0.98)' \
  --lr-scheduler polynomial_decay --lr 3e-05 --warmup-updates 2500 --total-num-update 40000 \
  --dropout 0.3 --attention-dropout 0.1 --weight-decay 0.0 \
  --max-tokens 1024 --update-freq 2 \
  --save-interval 1 --save-interval-updates 5000 --keep-interval-updates 1 --no-epoch-checkpoints \
  --seed 222 --log-format simple --log-interval 100 \
  --restore-file $PRETRAIN \
  --reset-optimizer --reset-meters --reset-dataloader --reset-lr-scheduler \
  --langs $langs \
  --ddp-backend no_c10d --skip-invalid-size-inputs-valid-test`

Out of curiosity, I also fine-tuned with the options switched (source as Arabic and target as English) and got translations from the checkpoint in reverse of expected order (i.e., English to Arabic). The command used for this fine-tuning was:

`fairseq-train <path_to_data> \
  --encoder-normalize-before --decoder-normalize-before \
  --arch mbart_large --layernorm-embedding \
  --task translation_from_pretrained_bart \
  --source-lang ar_AR --target-lang en_XX \
  --criterion label_smoothed_cross_entropy --label-smoothing 0.2 \
  --optimizer adam --adam-eps 1e-06 --adam-betas '(0.9, 0.98)' \
  --lr-scheduler polynomial_decay --lr 3e-05 --warmup-updates 2500 --total-num-update 40000 \
  --dropout 0.3 --attention-dropout 0.1 --weight-decay 0.0 \
  --max-tokens 1024 --update-freq 2 \
  --save-interval 1 --save-interval-updates 5000 --keep-interval-updates 1 --no-epoch-checkpoints \
  --seed 222 --log-format simple --log-interval 100 \
  --restore-file $PRETRAIN \
  --reset-optimizer --reset-meters --reset-dataloader --reset-lr-scheduler \
  --langs $langs \
  --ddp-backend no_c10d --skip-invalid-size-inputs-valid-test`

Attempting to translate from the expected source language returns the original text as given or sometimes the original text with contractions. The code for translation was essentially as suggested in the tutorial linked above (using the BART hub interface). I repeated the process (fine-tuning and translation) multiple times and preserved the commands I used but still got the same results, so it is highly unlikely that I messed up the commands as given in the official instructions. Any insight is appreciated. Thank you.

 - fairseq version: 0.10.1
 - PyTorch version: 1.7.1
 - OS: Linux (Ubuntu)
 - How you installed fairseq: conda
 - Python version: 3.8.8

**Comment by joao-alves97** [comment]
Hello! I'm having the same issue, how did you solve it? Thanks

**Comment by humamrashid** [comment]
Hi, in my case, the issue had to do with how the data was split before
fine-tuning rather than the dataset or the model. I don't know if you're
following the same tutorial to get it set up as I was at the time but this
is how I ended up splitting the data:

# Dataset (Ar-En) splitting code, adapted from:
#
https://tmramalho.github.io/science/2020/06/10/fine-tune-neural-translation-models-with-mBART/
# Dataset used is from WikiMatrix.

from tqdm import tqdm, trange
from sklearn.model_selection import train_test_split

ar_data = []
en_data = []

with open('WikiMatrix.ar-en.tsv') as fp:
    for line in tqdm(fp, total=1968009):
        line_data = line.rstrip().split('\t')
        ar_data.append(line_data[1] + '\n')
        en_data.append(line_data[2] + '\n')

total_test = 100000

ar_train, ar_subtotal, en_train, en_subtotal = train_test_split(
        ar_data, en_data, train_size=0.8, random_state=42)

ar_test, ar_val, en_test, en_val = train_test_split(
        ar_subtotal, en_subtotal, test_size=0.5, random_state=42)

file_mapping = {
    'train.ar_AR': ar_train,
    'train.en_XX': en_train,
    'valid.ar_AR': ar_val,
    'valid.en_XX': en_val,
    'test.ar_AR': ar_test,
    'test.en_XX': en_test,

}
for k, v in file_mapping.items():
    with open(f'enar_pre/{k}', 'w') as fp:
        fp.writelines(v)

# EOF.

Hope that helps,
Humam
